lib/cocina/models/builders/title_builder.rb

```diff
@@ -27,92 +27,127 @@ def self.build(titles, strategy: :first, add_punctuation: true)
+        def initialize(strategy:, add_punctuation:, only_one_parallel_value: true)
           @strategy = strategy
           @add_punctuation = add_punctuation
+          @only_one_parallel_value = only_one_parallel_value
         end
 
         # @param [[Array<Cocina::Models::Title>] cocina_titles the titles to consider
         # @return [String] the title value for Solr
+        #
+        # rubocop:disable Metrics/PerceivedComplexity
         def build(cocina_titles)
           cocina_title = primary_title(cocina_titles) || untyped_title(cocina_titles)
           cocina_title = other_title(cocina_titles) if cocina_title.blank?
 
           if strategy == :first
             extract_title(cocina_title)
           else
-            cocina_titles.map { |ctitle| extract_title(ctitle) }.flatten
+            result = cocina_titles.map { |ctitle| extract_title(ctitle) }.flatten
+            if only_one_parallel_value? && result.length == 1
+              result.first
+            else
+              result
+            end
           end
         end
+        # rubocop:enable Metrics/PerceivedComplexity
 
+        # @return [Array<String>] the main title value(s) for Solr - can be array due to parallel titles
         def main_title(titles)
           cocina_title = primary_title(titles) || untyped_title(titles)
           cocina_title = other_title(titles) if cocina_title.blank?
 
-          cocina_title = cocina_title.first if cocina_title.is_a?(Array)
           extract_main_title(cocina_title)
         end
 
         private
 
         attr_reader :strategy
 
+        # rubocop:disable Metrics/CyclomaticComplexity
+        # rubocop:disable Metrics/MethodLength
+        # rubocop:disable Metrics/PerceivedComplexity
         def extract_title(cocina_title)
+          title_values = if cocina_title.value
+                           cocina_title.value
+                         elsif cocina_title.structuredValue.present?
+                           rebuild_structured_value(cocina_title)
+                         elsif cocina_title.parallelValue.present?
+                           primary = cocina_title.parallelValue.find { |pvalue| pvalue.status == 'primary' }
+                           if primary && only_one_parallel_value? && strategy == :first
+                             extract_title(primary)
+                           elsif only_one_parallel_value? && strategy == :first
+                             untyped = cocina_title.parallelValue.find { |pvalue| pvalue.type.blank? }
+                             extract_title(untyped || cocina_title.parallelValue.first)
+                           else
+                             cocina_title.parallelValue.map { |pvalue| extract_title(pvalue) }
+                           end
+                         end
+          result = [title_values].flatten.compact.map { |val| remove_trailing_punctuation(val.strip) }
+          result.length == 1 ? result.first : result
```

Comment:
this got a lot uglier, and I know it ... however the specs pass and I'm kinda out of gas for refactoring at the moment.